### ISSUE TITLE
Support ES6 modules (import+export declarations)

### ISSUE THREAD

**Issue by cvrebert** [eslint/espree#35]
Refs #10. Esprima `harmony` already supports this.

**Comment by mlegenhausen** [comment]
+1

**Comment by nzakas** [comment]
ES6 progress is being tracked here: https://github.com/eslint/espree/issues/10

**Comment by nzakas** [comment]
Oops, hit enter too soon. We will be adding each ES6 feature from Esprima piece by piece. Module will likely be last unless someone submits a PR.

**Comment by cvrebert** [comment]
Well, there's not even a bullet point for it in #10's checklist currently.

**Comment by nzakas** [comment]
Thanks, I just added it.

**Comment by caridy** [comment]
@nzakas I can probably take care of the modules grammar as I did for esprima a while ago.

**Comment by Flickdm** [comment]
+1

**Comment by rjhilgefort** [comment]
:+1:

**Comment by rosshadden** [comment]
:+1:

**Comment by krivaten** [comment]
:+1:

**Comment by cvrebert** [comment]
:heart_eyes:

**Comment by nzakas** [comment]
@caridy awesome, we'd appreciate the contribution.

**Comment by caridy** [comment]
ok, I got a chance to put some time on this last night, and a questions arise:

In esprima, we introduced `{sourceType: "module"}`, which defines `extra.isModule` flag to signal to esprima that is should enforce `strict mode`, and should enable export/import declarations.

In espree, it seems that `extra.ecmaFeatures.<something>` will be the way to switch modules on. The problem is that enabling modules implies enabling other features as per the modules' grammar, e.g.: `import let foo = 1;`, which requires `blockBindings` to be available. Is there a provision to define dependencies between features? My proposal will be to add a provision similar to `{sourceType: "module"}`, which will flip multiple features under `extra.ecmaFeatures`, in this case, modules and `blockBindings`, and eventually classes.

**Comment by nzakas** [comment]
I think having `sourceType: "module"` is the right choice, both for Esprima compatibility and also because I don't like the idea of `ecmaFeatures` having dependencies on one another.

So +1 to your proposal.

**Comment by MoOx** [comment]
Interested too by import/export support in espree :)

**Comment by naholyr** [comment]
:+1:

**Comment by josiahsavary** [comment]
:thumbsup: This is awesome. When might can we expect it to get published to npm?

**Comment by nzakas** [comment]
When the work has been done. :)

I appreciate everyone's enthusiasm, but there's no need for additional +1ing. We're doing our best to get this done and this issue will be closed when the work is complete.

**Comment by naholyr** [comment]
You're awesome <3

**Comment by ai** [comment]
Does [eslint-babel](https://github.com/babel/babel-eslint) can help in this issue?

**Comment by douglasduteil** [comment]
+1

**Comment by kentcdodds** [comment]
@ai, eslint-babel doesn't support it properly either. Given:

```javascript
import foo as 'foo';

foo.bar();
```

currently results in:

```
3:1  error  'foo' is not defined                              no-undef
```
